File: server/speak.py
"""Speech as it is synthesised, for a voice agent that should not go silent.

The synthesiser returns a piece of text only once the whole of it is generated.
Waiting for the whole answer makes the listener wait for its last sentence
before hearing the first, so the text is cut into pieces and each piece is sent
as soon as it is ready. With Qwen3-TTS on an RTX 3080 synthesis runs at ×1.5
(engines/tts_fast.py), so after the first piece the next one is ready before
the current one finishes playing.

The cuts still go where a listener expects a pause. On a card slower than real
time there will be gaps between pieces; at sentence boundaries they sound like
breathing, in the middle of a phrase they sound like a stutter. So:

  - the first piece ends at the first clause boundary (a comma after three
    words or more) — the first sound is what the listener waits for;
  - every later piece ends at the end of a sentence;
  - a short sentence is joined to the next one when both are already known,
    because each call has a fixed cost;
  - a sentence longer than MAX_CHUNK is cut at its last clause boundary.

Two ways in:

  POST /v1/audio/speech with `stream_format` — OpenAI's parameter. "audio" sends
  the bytes as they are made (pcm, wav, mp3), "sse" sends `speech.audio.delta`
  events with base64 audio and ends with `speech.audio.done`.

  WS /v1/audio/speech/stream — for an agent that has the text only as its LLM
  produces it:

    client → {"type": "text", "text": "..."}   — more text, any size, even a token
    client → {"type": "flush"}                 — speak what is buffered now
    client → {"type": "cancel"}                — the user interrupted: drop everything
    client → {"type": "end"}                   — no more text: finish and close
    server → {"type": "ready", "sample_rate", "voice"}
    server → {"type": "synthesizing", "index", "text"}
    server → {"type": "audio", "index", "text", "seconds", "synth_seconds"}
             followed by one binary frame: PCM s16le mono at `sample_rate`
    server → {"type": "cancelled"}
    server → {"type": "done", "chunks", "seconds"}

Streaming does not go through the job queue — an agent's reply waiting behind a
long file would be no reply. It takes the synthesiser's lock piece by piece, so a
running synthesis job still delays it by the rest of that job.
"""
from __future__ import annotations

import asyncio
import base64
import json
import re
import struct
import time
from dataclasses import dataclass

import numpy as np
from fastapi import HTTPException, WebSocket, WebSocketDisconnect
from fastapi.responses import StreamingResponse

import audio_io
import textprep
import voices as voice_registry
from engines.base import Unsupported

MAX_CHUNK = 200             # символов — длиннее режем по запятой
MIN_CHUNK = 25              # символов — короче склеиваем с соседом, если он уже есть
FIRST_CLAUSE_WORDS = 3      # первый кусок можно отрезать по запятой после стольких слов
STREAM_FORMATS = ("pcm", "wav", "mp3")

# Конец предложения — знак, пробел и слово не со строчной буквы: так «т. е.»,
# «см. выше» и «3.5» не режутся. Поток ради этого ждёт первую букву следующего
# предложения — один токен.
SENTENCE = re.compile(r"[.!?…]+[\"»)\]]*(?=\s+[^\sa-zа-яё])|\n")
CLAUSE = re.compile(r"[,;:—–](?=\s)")


class Cancelled(Exception):
    pass


# --------------------------------------------------------------------- нарезка

class Segmenter:
    """Text in any pieces — tokens, lines, whole paragraphs; speakable chunks out."""

    def __init__(self):
        self.buf = ""
        self.emitted = 0            # кусков отдано наружу
        self.cuts = 0               # кусков отрезано, включая ещё не отданные

    def push(self, text: str) -> list[str]:
        self.buf += text
        return self._take(final=False)

    def flush(self) -> list[str]:
        return self._take(final=True)

    def split(self, text: str) -> list[str]:
        """The whole text at once: short sentences join their neighbours on both sides."""
        self.buf += text
        return self._take(final=True)

    def reset(self) -> None:
        self.buf, self.emitted, self.cuts = "", 0, 0

    def _cut(self, final: bool) -> str | None:
        m = SENTENCE.search(self.buf)
        if self.cuts == 0:
            # первый кусок — до первой запятой, если она раньше конца предложения
            for c in CLAUSE.finditer(self.buf, 0, m.start() if m else len(self.buf)):
                if len(self.buf[:c.start()].split()) >= FIRST_CLAUSE_WORDS:
                    return self._split(c.end())
        if m:
            return self._split(m.end())
        if len(self.buf) > MAX_CHUNK:
            head = self.buf[:MAX_CHUNK]
            cuts = [c.end() for c in CLAUSE.finditer(head)]
            return self._split(cuts[-1] if cuts else (head.rfind(" ") + 1 or MAX_CHUNK))
        if final and self.buf.strip():
            return self._split(len(self.buf))
        return None

    def _split(self, at: int) -> str:
        piece, self.buf = self.buf[:at], self.buf[at:]
        if piece.strip():
            self.cuts += 1
        return piece

    def _take(self, final: bool) -> list[str]:
        pieces = []
        while (p := self._cut(final)) is not None:
            if p.strip():
                pieces.append(p.strip())
        out: list[str] = []
        for p in pieces:
            # короткое склеиваем со следующим, раз оно уже есть; первый кусок
            # потока не трогаем — он ради того, чтобы звук начался раньше
            stream_first = self.emitted == 0 and len(out) == 1
            if out and len(out[-1]) < MIN_CHUNK and not stream_first:
                out[-1] = f"{out[-1]} {p}"
            else:
                out.append(p)
        self.emitted += len(out)
        return out


# ---------------------------------------------------------------------- синтез

@dataclass
class Options:
    voice: voice_registry.Voice
    language: str | None = None
    speed: float = 1.0
    prepare: bool = False
    legato: bool = False
    seed: int | None = None


def synthesize(tts, text: str, o: Options, cancelled) -> tuple[np.ndarray, int, float]:
    """One chunk. `cancelled()` is polled whenever the engine reports progress."""
    if o.prepare or o.legato:
        text = textprep.prepare(text, o.prepare, o.legato)["text"]

    def on_progress(_) -> None:
        if cancelled():
            raise Cancelled()

    started = time.perf_counter()
    out = tts.speak(text, o.voice.path, o.voice.text, language=o.language,
                    seed=o.seed, on_progress=on_progress)
    audio = audio_io.stretch(out.audio, out.sample_rate, o.speed) if o.speed != 1.0 else out.audio
    return audio, out.sample_rate, time.perf_counter() - started


def _wav_header(sr: int) -> bytes:
    """A wav header for a stream of unknown length: sizes set to the maximum,
    which players read as "until the data ends"."""
    unknown = 0xFFFFFFFF
    return (b"RIFF" + struct.pack("<I", unknown) + b"WAVE"
            + b"fmt " + struct.pack("<IHHIIHH", 16, 1, 1, sr, sr * 2, 2, 16)
            + b"data" + struct.pack("<I", unknown))


def pick_voice(tts, name: str | None) -> voice_registry.Voice:
    v = voice_registry.get(name) if name else voice_registry.default()
    if v is None:
        raise HTTPException(400, f"unknown voice '{name}'" if name else
                            "no voice available — add one via /v1/voices")
    if tts.needs_reference_text and not v.text:
        raise HTTPException(400, f"voice '{v.name}' has no reference transcript")
    return v


# ------------------------------------------------------------------------ HTTP

def http_stream(tts, req) -> StreamingResponse:
    """`/v1/audio/speech` with `stream_format` set."""
    fmt = (req.response_format or "wav").lower()
    if req.stream_format == "audio" and fmt not in STREAM_FORMATS:
        raise HTTPException(400, f"stream_format=audio supports {', '.join(STREAM_FORMATS)}; "
                                 f"use sse for {fmt}")
    if not req.input.strip():
        raise HTTPException(400, "input is empty")
    o = Options(voice=pick_voice(tts, req.voice), language=tts.language(req.language),
                speed=req.speed,
                prepare=req.prepare, legato=req.legato, seed=req.seed)
    seg = Segmenter()
    chunks = seg.split(req.input)
    state = {"stop": False}

    async def gen():
        total, t0 = 0.0, time.perf_counter()
        try:
            if req.stream_format == "audio" and fmt == "wav":
                yield _wav_header(tts.sample_rate)
            for text in chunks:
                audio, sr, _ = await asyncio.to_thread(
                    synthesize, tts, text, o, lambda: state["stop"])
                total += len(audio) / sr
                if req.stream_format == "audio":
                    if fmt in ("pcm", "wav"):
                        yield audio_io.to_pcm16(audio)
                    else:
                        yield (await asyncio.to_thread(audio_io.encode, audio, sr, fmt))[0]
                else:
                    data, _ = await asyncio.to_thread(audio_io.encode, audio, sr, fmt)
                    event = {"type": "speech.audio.delta", "text": text,
                             "audio": base64.b64encode(data).decode()}
                    yield f"data: {json.dumps(event, ensure_ascii=False)}\n\n"
            if req.stream_format == "sse":
                done = {"type": "speech.audio.done", "chunks": len(chunks),
                        "seconds": round(total, 2),
                        "generation_seconds": round(time.perf_counter() - t0, 2)}
                yield f"data: {json.dumps(done)}\n\n"
        finally:
            state["stop"] = True                     # клиент ушёл — дальше не синтезируем

    if req.stream_format == "sse":
        media = "text/event-stream"
    else:
        media = "audio/pcm" if fmt == "pcm" else audio_io.CONTENT_TYPES[fmt]
    return StreamingResponse(gen(), media_type=media, headers={
        "X-Voice": o.voice.name, "X-Chunks": str(len(chunks)),
        "X-Sample-Rate": str(tts.sample_rate), "Cache-Control": "no-cache",
        "X-Accel-Buffering": "no"})


# ------------------------------------------------------------------- WebSocket

def attach(app, tts):
    @app.websocket("/v1/audio/speech/stream")
    async def speak(ws: WebSocket, voice: str | None = None, language: str | None = None,
                    speed: float = 1.0, sample_rate: int | None = None,
                    prepare: bool = False, legato: bool = False, seed: int | None = None):
        await ws.accept()
        sample_rate = sample_rate or tts.sample_rate
        try:
            if not 8000 <= sample_rate <= 48000:
                raise HTTPException(400, "sample_rate must be 8000–48000")
            o = Options(voice=pick_voice(tts, voice), language=tts.language(language),
                        speed=min(1.2, max(0.8, speed)) if speed != 1.0 else 1.0,
                        prepare=prepare, legato=legato, seed=seed)
        except (HTTPException, Unsupported) as e:
            await ws.send_json({"type": "error",
                                "error": e.detail if isinstance(e, HTTPException) else str(e)})
            await ws.close(code=1003)
            return

        seg = Segmenter()
        todo: asyncio.Queue = asyncio.Queue()
        epoch = {"n": 0}                 # отмена увеличивает эпоху; старые куски выбрасываются
        send_lock = asyncio.Lock()
        END = object()

        async def send_json(event: dict) -> None:
            async with send_lock:
                await ws.send_json(event)

        async def reader() -> None:
            while True:
                msg = await ws.receive()
                if msg["type"] == "websocket.disconnect":
                    epoch["n"] += 1
                    await todo.put(None)
                    return
                raw = msg.get("text")
                if raw is None:
                    continue
                try:
                    cmd = json.loads(raw)
                except ValueError:
                    cmd = {"type": "text", "text": raw}   # просто текст — тоже текст
                kind = cmd.get("type")
                if kind == "text":
                    for p in seg.push(str(cmd.get("text", ""))):
                        await todo.put((epoch["n"], p))
                elif kind == "flush":
                    for p in seg.flush():
                        await todo.put((epoch["n"], p))
                elif kind == "cancel":
                    epoch["n"] += 1
                    seg.reset()
                    while not todo.empty():
                        todo.get_nowait()
                    await send_json({"type": "cancelled"})
                elif kind == "end":
                    for p in seg.flush():
                        await todo.put((epoch["n"], p))
                    await todo.put(END)
                    return

        await send_json({"type": "ready", "sample_rate": sample_rate, "voice": o.voice.name})
        reading = asyncio.create_task(reader())
        index, seconds = 0, 0.0
        try:
            while True:
                item = await todo.get()
                if item is None:
                    return                               # клиент ушёл
                if item is END:
                    await send_json({"type": "done", "chunks": index,
                                     "seconds": round(seconds, 2)})
                    await ws.close()
                    return
                mine, text = item
                if mine != epoch["n"]:
                    continue
                await send_json({"type": "synthesizing", "index": index, "text": text})
                try:
                    audio, sr, took = await asyncio.to_thread(
                        synthesize, tts, text, o, lambda: mine != epoch["n"])
                except Cancelled:
                    continue
                if mine != epoch["n"]:
                    continue                             # отменили, пока кодировали
                if sample_rate != sr:
                    import soxr
                    audio = soxr.resample(audio, sr, sample_rate).astype(np.float32)
                dur = len(audio) / sample_rate
                seconds += dur
                async with send_lock:
                    await ws.send_json({"type": "audio", "index": index, "text": text,
                                        "seconds": round(dur, 2),
                                        "synth_seconds": round(took, 2)})
                    await ws.send_bytes(audio_io.to_pcm16(audio))
                index += 1
        except WebSocketDisconnect:
            pass
        except Exception as e:                           # noqa: BLE001 — уходит клиенту
            try:
                await send_json({"type": "error", "error": str(e)})
                await ws.close(code=1011)
            except Exception:                            # noqa: BLE001
                pass
        finally:
            epoch["n"] += 1
            reading.cancel()

File: rust/src/speak.rs
//! Speech as it is synthesised, for a voice agent that should not go silent —
//! server/speak.py.
//!
//! The synthesiser returns a piece of text only once the whole of it is
//! generated, so the text is cut into pieces and each is sent as soon as it is
//! ready. The cuts go where a listener expects a pause: the first piece at the
//! first clause boundary (the first sound is what the listener waits for), the
//! rest at sentence ends; short sentences are joined, long ones cut at their
//! last clause boundary.
//!
//!   POST /v1/audio/speech with `stream_format` — "audio" sends bytes as they
//!   are made (pcm, wav, mp3), "sse" sends `speech.audio.delta` events.
//!
//!   WS /v1/audio/speech/stream — text in as an LLM writes it, audio out in
//!   pieces, cancel when the user interrupts.

use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::{Arc, OnceLock};
use std::time::Instant;

use axum::body::Body;
use axum::extract::ws::{Message, WebSocket};
use axum::http::header;
use axum::response::{IntoResponse, Response};
use base64::Engine as _;
use fancy_regex::Regex;
use serde::Deserialize;
use serde_json::{Value, json};
use tokio::sync::mpsc;

use crate::App;
use crate::audio;
use crate::engines::Stop;
use crate::errors::{ApiError, ApiResult};
use crate::jobs::round;
use crate::textprep;
use crate::voices::Voice;

const MAX_CHUNK: usize = 200; // символов — длиннее режем по запятой
const MIN_CHUNK: usize = 25; // символов — короче склеиваем с соседом, если он уже есть
const FIRST_CLAUSE_WORDS: usize = 3; // первый кусок можно отрезать по запятой после стольких слов
pub const STREAM_FORMATS: [&str; 3] = ["pcm", "wav", "mp3"];

fn sentence() -> &'static Regex {
    // Конец предложения — знак, пробел и слово не со строчной буквы: так «т. е.»,
    // «см. выше» и «3.5» не режутся.
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r#"[.!?…]+["»)\]]*(?=\s+[^\sa-zа-яё])|\n"#).unwrap())
}

fn clause() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r"[,;:—–](?=\s)").unwrap())
}

fn byte_at(s: &str, chars: usize) -> usize {
    s.char_indices().nth(chars).map_or(s.len(), |(i, _)| i)
}

// --------------------------------------------------------------------- нарезка

/// Text in any pieces — tokens, lines, whole paragraphs; speakable chunks out.
#[derive(Default)]
pub struct Segmenter {
    buf: String,
    emitted: usize, // кусков отдано наружу
    cuts: usize,    // кусков отрезано, включая ещё не отданные
}

impl Segmenter {
    pub fn push(&mut self, text: &str) -> Vec<String> {
        self.buf.push_str(text);
        self.take(false)
    }

    pub fn flush(&mut self) -> Vec<String> {
        self.take(true)
    }

    /// The whole text at once: short sentences join their neighbours on both sides.
    pub fn split(&mut self, text: &str) -> Vec<String> {
        self.buf.push_str(text);
        self.take(true)
    }

    pub fn reset(&mut self) {
        *self = Segmenter::default();
    }

    fn cut(&mut self, fin: bool) -> Option<String> {
        let m = sentence().find(&self.buf).ok().flatten().map(|m| (m.start(), m.end()));
        if self.cuts == 0 {
            // первый кусок — до первой запятой, если она раньше конца предложения
            let end = m.map_or(self.buf.len(), |(s, _)| s);
            let region = &self.buf[..end];
            let at = clause()
                .find_iter(region)
                .flatten()
                .find(|c| region[..c.start()].split_whitespace().count() >= FIRST_CLAUSE_WORDS)
                .map(|c| c.end());
            if let Some(at) = at {
                return Some(self.split_at(at));
            }
        }
        if let Some((_, end)) = m {
            return Some(self.split_at(end));
        }
        if self.buf.chars().count() > MAX_CHUNK {
            let head_end = byte_at(&self.buf, MAX_CHUNK);
            let head = &self.buf[..head_end];
            let at = clause()
                .find_iter(head)
                .flatten()
                .last()
                .map(|c| c.end())
                .or_else(|| head.rfind(' ').map(|i| i + 1))
                .filter(|&i| i > 0)
                .unwrap_or(head_end);
            return Some(self.split_at(at));
        }
        if fin && !self.buf.trim().is_empty() {
            return Some(self.split_at(self.buf.len()));
        }
        None
    }

    fn split_at(&mut self, at: usize) -> String {
        let rest = self.buf.split_off(at);
        let piece = std::mem::replace(&mut self.buf, rest);
        if !piece.trim().is_empty() {
            self.cuts += 1;
        }
        piece
    }

    fn take(&mut self, fin: bool) -> Vec<String> {
        let mut pieces = vec![];
        while let Some(p) = self.cut(fin) {
            if !p.trim().is_empty() {
                pieces.push(p.trim().to_string());
            }
        }
        let mut out: Vec<String> = vec![];
        for p in pieces {
            // короткое склеиваем со следующим, раз оно уже есть; первый кусок
            // потока не трогаем — он ради того, чтобы звук начался раньше
            let stream_first = self.emitted == 0 && out.len() == 1;
            match out.last_mut() {
                Some(last) if last.chars().count() < MIN_CHUNK && !stream_first => {
                    last.push(' ');
                    last.push_str(&p);
                }
                _ => out.push(p),
            }
        }
        self.emitted += out.len();
        out
    }
}

// ---------------------------------------------------------------------- синтез

#[derive(Clone)]
pub struct Options {
    pub voice: Voice,
    pub language: String,
    pub speed: f64,
    pub prepare: bool,
    pub legato: bool,
    /// знаки ударения; решено заранее: запрошены, язык русский, модель их понимает
    pub stress: bool,
    pub seed: Option<i64>,
}

pub fn pick_voice(app: &App, name: Option<&str>) -> ApiResult<Voice> {
    let v = match name.filter(|n| !n.is_empty()) {
        Some(n) => app.voices.get(n).ok_or_else(|| ApiError::bad(format!("unknown voice '{n}'")))?,
        None => app.voices.default().ok_or_else(|| ApiError::bad("no voice available — add one via /v1/voices"))?,
    };
    if app.engines.needs_reference_text() && v.text.is_empty() {
        return Err(ApiError::bad(format!("voice '{}' has no reference transcript", v.name)));
    }
    Ok(v)
}

pub fn speak_args(text: &str, o: &Options) -> Value {
    json!({"text": text, "ref_audio": o.voice.path.to_string_lossy(), "ref_text": o.voice.text,
           "language": o.language, "seed": o.seed})
}

/// One chunk. `alive()` is asked at every progress event; false stops the work.
pub async fn synthesize(app: &App, text: &str, o: &Options, alive: impl Fn() -> bool)
                        -> Result<(Vec<f32>, u32, f64), ApiError> {
    let text = textprep::for_speech(text, o.prepare, o.legato, o.stress).await;
    let started = Instant::now();
    let out = app.engines.speak(speak_args(&text, o), |_| alive()).await.map_err(|s| match s {
        Stop::Cancelled => ApiError::new(409, "cancelled"),
        Stop::Failed(e) => e.into(),
    })?;
    let audio = audio::stretch(out.audio, out.sample_rate, o.speed).await?;
    Ok((audio, out.sample_rate, started.elapsed().as_secs_f64()))
}

// ------------------------------------------------------------------------ HTTP

pub async fn http_stream(app: Arc<App>, req: crate::api::SpeechRequest) -> ApiResult<Response> {
    let fmt = req.response_format.to_lowercase();
    let sse = req.stream_format.as_deref() == Some("sse");
    if !sse && !STREAM_FORMATS.contains(&fmt.as_str()) {
        return Err(ApiError::bad(format!(
            "stream_format=audio supports {}; use sse for {fmt}", STREAM_FORMATS.join(", "))));
    }
    if req.input.trim().is_empty() {
        return Err(ApiError::bad("input is empty"));
    }
    let language = app.engines.language(req.language.as_deref()).await?;
    let o = Options {
        voice: pick_voice(&app, req.voice.as_deref())?,
        stress: req.stress && language == "ru" && app.engines.stress_marks(),
        language,
        speed: req.speed,
        prepare: req.prepare,
        legato: req.legato,
        seed: req.seed,
    };
    let chunks = Segmenter::default().split(&req.input);
    let (tx, rx) = mpsc::channel::<Result<Vec<u8>, std::io::Error>>(4);
    let rate = app.engines.tts_rate();
    let (voice, n) = (o.voice.name.clone(), chunks.len());
    let fmt2 = fmt.clone();
    tokio::spawn(async move {
        let (mut total, t0) = (0.0, Instant::now());
        if !sse && fmt2 == "wav" && tx.send(Ok(audio::wav_stream_header(rate))).await.is_err() {
            return;
        }
        for text in &chunks {
            // клиент ушёл — дальше не синтезируем: отмена уходит в движок
            let (a, sr, _) = match synthesize(&app, text, &o, || !tx.is_closed()).await {
                Ok(v) => v,
                Err(e) => {
                    let _ = tx.send(Err(std::io::Error::other(e.message))).await;
                    return;
                }
            };
            total += a.len() as f64 / sr as f64;
            let data = if !sse && (fmt2 == "pcm" || fmt2 == "wav") {
                audio::to_pcm16(&a)
            } else {
                match audio::encode(&a, sr, &fmt2).await {
                    Ok((d, _)) => d,
                    Err(e) => {
                        let _ = tx.send(Err(std::io::Error::other(e.message))).await;
                        return;
                    }
                }
            };
            let item = if sse {
                let ev = json!({"type": "speech.audio.delta", "text": text,
                                "audio": base64::engine::general_purpose::STANDARD.encode(&data)});
                format!("data: {ev}\n\n").into_bytes()
            } else {
                data
            };
            if tx.send(Ok(item)).await.is_err() {
                return;
            }
        }
        if sse {
            let done = json!({"type": "speech.audio.done", "chunks": chunks.len(), "seconds": round(total, 2),
                              "generation_seconds": round(t0.elapsed().as_secs_f64(), 2)});
            let _ = tx.send(Ok(format!("data: {done}\n\n").into_bytes())).await;
        }
    });
    let media = if sse { "text/event-stream" } else if fmt == "pcm" { "audio/pcm" } else { audio::content_type(&fmt) };
    let body = Body::from_stream(tokio_stream::wrappers::ReceiverStream::new(rx));
    Ok((
        [
            (header::CONTENT_TYPE, media.to_string()),
            (header::CACHE_CONTROL, "no-cache".into()),
            (header::HeaderName::from_static("x-voice"), voice),
            (header::HeaderName::from_static("x-chunks"), n.to_string()),
            (header::HeaderName::from_static("x-sample-rate"), rate.to_string()),
            (header::HeaderName::from_static("x-accel-buffering"), "no".into()),
        ],
        body,
    )
        .into_response())
}

// ------------------------------------------------------------------- WebSocket

#[derive(Deserialize)]
pub struct StreamQuery {
    voice: Option<String>,
    language: Option<String>,
    #[serde(default = "one")]
    speed: f64,
    sample_rate: Option<u32>,
    #[serde(default)]
    prepare: bool,
    #[serde(default)]
    legato: bool,
    #[serde(default = "yes")]
    stress: bool,
    seed: Option<i64>,
}

fn one() -> f64 {
    1.0
}

fn yes() -> bool {
    true
}

enum Out {
    Json(Value),
    Audio(Value, Vec<u8>),
    Close(Option<u16>),
}

enum Todo {
    Text(u64, String),
    End,
    Gone,
}

pub async fn socket(app: Arc<App>, q: StreamQuery, ws: WebSocket) {
    use futures::{SinkExt, StreamExt};
    let (mut sink, mut stream) = ws.split();
    let (out_tx, mut out_rx) = mpsc::unbounded_channel::<Out>();
    // один писатель — события и звук не перемешиваются
    let writer = tokio::spawn(async move {
        while let Some(o) = out_rx.recv().await {
            let r = match o {
                Out::Json(v) => sink.send(Message::Text(v.to_string().into())).await,
                Out::Audio(v, pcm) => match sink.send(Message::Text(v.to_string().into())).await {
                    Ok(()) => sink.send(Message::Binary(pcm.into())).await,
                    e => e,
                },
                Out::Close(code) => {
                    let frame = code.map(|c| axum::extract::ws::CloseFrame { code: c, reason: "".into() });
                    let _ = sink.send(Message::Close(frame)).await;
                    break;
                }
            };
            if r.is_err() {
                break;
            }
        }
    });
    let send = |v: Value| {
        let _ = out_tx.send(Out::Json(v));
    };

    let rate = q.sample_rate.unwrap_or_else(|| app.engines.tts_rate());
    let options = async {
        if !(8000..=48000).contains(&rate) {
            return Err(ApiError::bad("sample_rate must be 8000–48000"));
        }
        let language = app.engines.language(q.language.as_deref()).await?;
        Ok(Options {
            voice: pick_voice(&app, q.voice.as_deref())?,
            stress: q.stress && language == "ru" && app.engines.stress_marks(),
            language,
            speed: if q.speed != 1.0 { q.speed.clamp(0.8, 1.2) } else { 1.0 },
            prepare: q.prepare,
            legato: q.legato,
            seed: q.seed,
        })
    }
    .await;
    let o = match options {
        Ok(o) => o,
        Err(e) => {
            send(json!({"type": "error", "error": e.message}));
            let _ = out_tx.send(Out::Close(Some(1003)));
            let _ = writer.await;
            return;
        }
    };

    let epoch = Arc::new(AtomicU64::new(0)); // отмена увеличивает эпоху; старые куски выбрасываются
    let (todo_tx, mut todo_rx) = mpsc::unbounded_channel::<Todo>();
    send(json!({"type": "ready", "sample_rate": rate, "voice": o.voice.name}));

    let reader = {
        let (epoch, todo_tx, out_tx) = (epoch.clone(), todo_tx.clone(), out_tx.clone());
        tokio::spawn(async move {
            let mut seg = Segmenter::default();
            let put = |pieces: Vec<String>| {
                for p in pieces {
                    let _ = todo_tx.send(Todo::Text(epoch.load(Ordering::SeqCst), p));
                }
            };
            while let Some(msg) = stream.next().await {
                let raw = match msg {
                    Ok(Message::Text(t)) => t.to_string(),
                    Ok(Message::Close(_)) | Err(_) => break,
                    _ => continue,
                };
                // просто текст — тоже текст
                let cmd: Value = serde_json::from_str(&raw).unwrap_or_else(|_| json!({"type": "text", "text": raw}));
                match cmd["type"].as_str() {
                    Some("text") => {
                        let t = match &cmd["text"] {
                            Value::String(s) => s.clone(),
                            Value::Null => String::new(),
                            v => v.to_string(),
                        };
                        put(seg.push(&t));
                    }
                    Some("flush") => put(seg.flush()),
                    Some("cancel") => {
                        epoch.fetch_add(1, Ordering::SeqCst);
                        seg.reset();
                        let _ = out_tx.send(Out::Json(json!({"type": "cancelled"})));
                    }
                    Some("end") => {
                        put(seg.flush());
                        let _ = todo_tx.send(Todo::End);
                        return;
                    }
                    _ => {}
                }
            }
            epoch.fetch_add(1, Ordering::SeqCst);
            let _ = todo_tx.send(Todo::Gone);
        })
    };
    drop(todo_tx);

    let (mut index, mut seconds) = (0usize, 0.0f64);
    while let Some(item) = todo_rx.recv().await {
        let (mine, text) = match item {
            Todo::Gone => break,
            Todo::End => {
                send(json!({"type": "done", "chunks": index, "seconds": round(seconds, 2)}));
                let _ = out_tx.send(Out::Close(None));
                break;
            }
            Todo::Text(e, t) => (e, t),
        };
        if mine != epoch.load(Ordering::SeqCst) {
            continue;
        }
        send(json!({"type": "synthesizing", "index": index, "text": text}));
        let ep = epoch.clone();
        let (a, sr, took) = match synthesize(&app, &text, &o, || ep.load(Ordering::SeqCst) == mine).await {
            Ok(v) => v,
            Err(e) if e.status.as_u16() == 409 => continue,
            Err(e) => {
                send(json!({"type": "error", "error": e.message}));
                let _ = out_tx.send(Out::Close(Some(1011)));
                break;
            }
        };
        if mine != epoch.load(Ordering::SeqCst) {
            continue; // отменили, пока кодировали
        }
        let a = if rate != sr { audio::Resampler::whole(sr, rate, &a) } else { a };
        let dur = a.len() as f64 / rate as f64;
        seconds += dur;
        let ev = json!({"type": "audio", "index": index, "text": text, "seconds": round(dur, 2),
                        "synth_seconds": round(took, 2)});
        let _ = out_tx.send(Out::Audio(ev, audio::to_pcm16(&a)));
        index += 1;
    }
    epoch.fetch_add(1, Ordering::SeqCst);
    reader.abort();
    drop(out_tx);
    let _ = writer.await;
}
